File: quanta_tissu/tisslm/js/transformer/FeedForward.js
const { Matrix } = require('../core/Matrix.js');

class FeedForward {
    constructor(embedSize, hiddenSize) {
        this.embedSize = embedSize;
        this.hiddenSize = hiddenSize;

        // Linear layers
        this.w1 = Matrix.random(embedSize, hiddenSize);
        this.b1 = Matrix.zeros(1, hiddenSize);
        this.w2 = Matrix.random(hiddenSize, embedSize);
        this.b2 = Matrix.zeros(1, embedSize);
    }

    forward(x) {
        // x is a Matrix with shape (seq_len, embed_size)

        // First linear layer + ReLU
        let output = x.dot(this.w1).add(this.b1);
        output = output.relu();

        // Second linear layer
        output = output.dot(this.w2).add(this.b2);

        return output;
    }
}

module.exports = { FeedForward };

File: quanta_tissu/tisslm/program/cllm/feed_forward.cpp
#include "feed_forward.h"
#include <iostream>

namespace cllm {

FeedForward::FeedForward(const ModelConfig& config) : config_(config) {
    // A common choice for the hidden dimension is 4 * d_model.
    hidden_dim = 4 * config.d_model;

    // Initialize weights and biases.
    // In a real implementation, these would be loaded from a checkpoint.
    // For now, we'll initialize them with random values.
    weight1_ = Eigen::MatrixXf::Random(hidden_dim, config.d_model);
    bias1_ = Eigen::VectorXf::Random(hidden_dim);
    weight2_ = Eigen::MatrixXf::Random(config.d_model, hidden_dim);
    bias2_ = Eigen::VectorXf::Random(config.d_model);
}

// Performs the forward pass for the feed-forward network.
Eigen::MatrixXf FeedForward::forward(const Eigen::MatrixXf& input) {
    // First linear transformation
    Eigen::MatrixXf hidden = (input * weight1_.transpose()).rowwise() + bias1_.transpose();

    // ReLU activation function (element-wise)
    hidden = hidden.cwiseMax(0);

    // Second linear transformation
    Eigen::MatrixXf output = (hidden * weight2_.transpose()).rowwise() + bias2_.transpose();

    return output;
}

Eigen::MatrixXf FeedForward::backward(const Eigen::MatrixXf& d_out, const Eigen::MatrixXf& x, float lr) {
    Eigen::MatrixXf z = (x * weight1_.transpose()).rowwise() + bias1_.transpose();
    Eigen::MatrixXf h = z.cwiseMax(0);

    Eigen::MatrixXf d_weight2 = d_out.transpose() * h;
    Eigen::VectorXf d_bias2 = d_out.colwise().sum();
    Eigen::MatrixXf d_h = d_out * weight2_;

    Eigen::MatrixXf d_z = (z.array() > 0).cast<float>().matrix().cwiseProduct(d_h);
    Eigen::MatrixXf d_weight1 = d_z.transpose() * x;
    Eigen::VectorXf d_bias1 = d_z.colwise().sum();

    weight1_ -= lr * d_weight1;
    bias1_ -= lr * d_bias1;
    weight2_ -= lr * d_weight2;
    bias2_ -= lr * d_bias2;

    return d_z * weight1_;
}

} // namespace cllm
